src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.RestServer
+{
+    public interface IEgressOperationStore
+    {
+        Task<Guid> AddOperation(IEgressOperation streamResult);
+        Task CancelOperation(Guid operationId);
+        Task CompleteOperation(Guid operationId, ExecutionResult<EgressResult?> result);
+
+        //CONSIDER We are using the same object for internal store representation as we are for
+        //external controller responses.
+        Task<Models.OperationStatusBase> GetOperationStatus(Guid operationId);
+    }
+
+    internal sealed class EgressOperationStore : IEgressOperationStore
+    {
+        private sealed class EgressEntry
+        {
+            public ExecutionResult<EgressResult?> ExecutionResult { get; set; }
+            public Models.OperationState State { get; set;}
+
+            public EgressRequest EgressRequest { get; set; }
+
+            public DateTime CreatedDateTime { get; } = DateTime.UtcNow;
+
+            public Guid OperationId { get; set; }
+        }
+
+        private readonly Dictionary<Guid, EgressEntry> _requests = new();
+        private readonly IEgressOperationQueue _taskQueue;
+
+        public EgressOperationStore(IEgressOperationQueue queue)
+        {
+            _taskQueue = queue;
+        }
+
+        public async Task<Guid> AddOperation(IEgressOperation egressOperation)
+        {
+            Guid operationId = Guid.NewGuid();
+
+            var request = new EgressRequest(operationId, egressOperation);
+            lock (_requests)
+            {
+                //Add operation object to central table.
+                _requests.Add(operationId,
+                    new EgressEntry
+                    {
+                        State = Models.OperationState.Running,
+                        EgressRequest = request,
+                        OperationId = operationId
+                    });
+            }
+
+            //Kick off work to attempt egress
+            await _taskQueue.QueueBackgroundWorkItemAsync(request);
+
+            return operationId;
+        }
+
+        public Task CancelOperation(Guid operationId)
```

Comment:
This and the below methods should not be Task-returning methods if they aren't async, if the service is just an implementation within the server assembly.